### PR TITLE
Avoid unnecessary persistence of activity enablements #1084

### PR DESCRIPTION
**Pull request by HeikoKlare**
The `ActivityPersistenceHelper` persists all activity enablements upon every notification about a changed activity enablement it receives. Persistence is, however, only done and thus necessary for activities that are not expression-controlled. So the current implementation performs unnecessary persistence operations.

With this change, the ActivityPersistenceHelper only persists the enablements when the enablement of a non-expression-controlled activity has changed. To this end, the received ActivityManagerEvent is extended by the information about whether such an activity was affected. This information is either directly passed to the event construction in case it has already been calculated in the calling context or is calculated lazily when accessing the value.

Contributes to https://github.com/eclipse-platform/eclipse.platform.ui/issues/1084.

### CODE REVIEW

**Comment by HeikoKlare** [comment]
Failing test is documented: #926

File: bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/activities/ActivityManagerEvent.java [hunk 1]
```diff
@@ -214,4 +302,34 @@ public boolean haveDefinedCategoryIdsChanged() {
...
+
+	private static boolean haveEnabledNonExpressionControlledActivityIdsChanged(IActivityManager activityManager,
+			Set<String> previousIds) {
+		Set<String> currentIds = activityManager.getEnabledActivityIds();
+		Set<String> unchangedIds = new HashSet<>(currentIds);
+		unchangedIds.retainAll(previousIds);
+		Set<String> changedIds = new HashSet<>(previousIds);
+		changedIds.addAll(currentIds);
+		changedIds.removeAll(unchangedIds);
+		return changedIds.stream().anyMatch(id -> activityManager.getActivity(id).getExpression() == null);
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/activities/ActivityManagerEvent.java]: You could avoid creating new `Set`s if you do it like this:

```java
Set<String> currentIds = activityManager.getEnabledActivityIds();

Stream.concat(currentIds.stream(), previousIds.stream()) // the sum...
		.filter(id -> !currentIds.contains(id) || !previousIds.contains(id))// ...minus the intersection
		.anyMatch(id -> activityManager.getActivity(id).getExpression() == null);
```

You could even experiment concatenating 2 `.parallelStream()`s and see if it improves/degrades performance.

Comment (HeikoKlare) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/activities/ActivityManagerEvent.java]: Thanks! Good point. It should be even more efficient to treat the streams individually, as then only one `contains` checks for every element is necessary:
```java
	Set<String> currentIds = activityManager.getEnabledActivityIds();
	Stream<String> addedIds = currentIds.stream().filter(id -> !previousIds.contains(id));
	if (addedIds.anyMatch(this::isIdOfNonExpressionControlledActivity)) {
		return true;
	}
	Stream<String> removedIds = previousIds.stream().filter(id -> !currentIds.contains(id));
	return removedIds.anyMatch(this::isIdOfNonExpressionControlledActivity);

...

private boolean isIdOfNonExpressionControlledActivity(String id) {
	return activityManager.getActivity(id).getExpression() == null;
}
```
Actually, this method only exists for compatibility reasons. The actual goal is that the field `enabledNonExpressionControlledActivityIdsChanged` is set via the constructor call as the calculation has implicitly been done before. Thus, in the demo scenario changes to this method have no performance impact. Even if changing the code to explicitly performing the calculation everytime the value is accessed has no measurable performance impact, as the performance issue addressed by this PR does not come from this calculation but from the code that is executed when not performing this check at all.

I have updated the functionality in b8c6a1e accordingly.

**Comment by akurtakov** [comment]
Looks like this one broke javadoc generation https://download.eclipse.org/eclipse/downloads/drops4/I20230919-1800/compilelogs/platform.doc.isv.javadoc.txt

**Comment by laeubi** [comment]
Also reported here (search for `javadoc:jar`):

https://github.com/eclipse-platform/eclipse.platform.ui/runs/16931217956

**Comment by HeikoKlare** [comment]
Sorry for that. I'll provide a fix in a moment.

Could we have those checks integrated into mandatory GH Actions checks to have them represented as actual errors of the PR? Or is there a (non-technical) reason for not dong so?

**Comment by HeikoKlare** [comment]
> Also reported here (search for `javadoc:jar`):
> 
> https://github.com/eclipse-platform/eclipse.platform.ui/runs/16931217956

I think the ones reported there are not related to this PR but pre-existing issues in Javadoc generation, as they appear in every Jenkins log. In particular, the warnings also occur in the build for the PR with the fixes for the broken Javadoc in this PR: https://github.com/eclipse-platform/eclipse.platform.ui/pull/1142/checks?check_run_id=16954953767

**Comment by akurtakov** [comment]
The I-build javadoc generation cares about public API only while the GH action looks at all the sources.  We never managed to get that to a clean state to even consider making it fail the build. It's an extra issue (not investigated yet AFAICT) that some javadoc errors are not catched in the Javadoc report but are only in the Maven one which is full with other issues and thus even less people notice these. 
Long story short - it's a a web of issues so I kindly ask everyone to just try to not let warnings stay around when noticed, if enough people do that at some point we should be able to raise certain things to errors and make them more obvious. Help on the releng side to get better configs, improving tools and etc. is also highly appreciated.

**Comment by HeikoKlare** [comment]
Thanks for clarification, @akurtakov! Good to know about that situation and have it in mind when looking at PR check results and logs.
